src/batch.jl

```diff
@@ -5,7 +5,8 @@ function (b::BatchClosure{F,A,B})(p::Ptr{UInt}) where {F,A,B}
   (offset, args) = ThreadingUtilities.load(p, A, 2*sizeof(UInt))
   (offset, start) = ThreadingUtilities.load(p, UInt, offset)
   (offset, stop ) = ThreadingUtilities.load(p, UInt, offset)
-  b.f(args, (start+one(UInt))%Int, stop%Int)
+  (offset, i ) = ThreadingUtilities.load(p, UInt, offset)
+  b.f(args, (start+one(UInt))%Int, stop%Int, i%Int)
```

Comment:
Now `i`, the enumeration index of the thread is used and provided as an argument to the closure.

Reply:
An alternative with lower overhead would be to
1) Require users to provide a type
2) Have the closure return the tasklocal variables instead of storing them into the vector.
3) Add a `(::BatchClosure{F,A,B})` method, perhaps with a type parameter `Tuple{...}` giving the types of all tasklocals (`Tuple{}` given none), then have it store each of these in the tasklocal pointer.
4) the `f!` not in a loop can store to the end of the tasklocal vector.
5) the loop over all thread tasks that calls `wait` can then load and store them into the vector.

You could then eventually add alternatives to creating and storing a vector, e.g. maybe someone wants to sum these instead.

If this sounds like too much, I'd like to at least avoid storing and loading `i` unless we actually have tasklocal variables.